
// Generated by ntoolkit/futures
package futures_test

import "ntoolkit/futures"

import "testing"


type DeferredT struct {
	DeferredValue futures.Promise
}

func (promise *DeferredT) init() {
	if promise.DeferredValue == nil {
		promise.DeferredValue = &futures.DeferredValue{}
	}
}

func (promise *DeferredT) Resolve(result *testing.T) {
	promise.init()
	promise.DeferredValue.PResolve(result)
}

func (promise *DeferredT) Reject(err error) {
	promise.init()
	promise.DeferredValue.PReject(err)
}

func (promise *DeferredT) Errors() []error {
	promise.init()
	return promise.DeferredValue.PErrors()
}

func (promise *DeferredT) Then(resolve func(*testing.T), reject func(error)) *DeferredT {
	promise.init()
	promise.DeferredValue.PThen(func(value interface{}) {
		if v, ok := value.(*testing.T); ok {
			resolve(v)
		} else {
		  panic("Invalid value used to resolve DeferredT")
		}
	}, reject)
	return promise
}

func (promise *DeferredT) PThen(result func(interface{}), reject func(error)) futures.Promise {
	promise.init()
	return promise.DeferredValue.PThen(result, reject)
}

func (promise *DeferredT) PErrors() []error {
	promise.init()
	return promise.DeferredValue.PErrors()
}

func (promise *DeferredT) PResolve(result interface{}) {
	promise.init()
	promise.DeferredValue.PResolve(result)
}

func (promise *DeferredT) PReject(err error) {
	promise.init()
	promise.DeferredValue.PReject(err)
}
